text/0133-explicit-layer-order.md

```diff
@@ -0,0 +1,226 @@
+
+> - The lifecycle MUST order all `<layer>` paths to reflect the reversed order of the buildpack group.
+> - The lifecycle MUST order all `<layer>` paths provided by a given buildpack alphabetically ascending.
+
+While multiple buildpacks have their execution order preserved implicitly, the layer ordering within a buildpack depends on the layer's name. So changing the name of the layers `gems` and `ruby` to something like `binruby` and `user_gems` would have the effect that executables such as `rake` (which ships with ruby) might now be loaded from a different layer path because the name has a different alphanumeric ordering.
+
+The ordering of layers is important, and the current ordering mechanism (alphanumeric layer names) is not intuitive.
+
+Layers need to be loaded in several locations:
+
+- At "build" time, multiple buildpacks can run simultaneously. If a layer is created with `build=true` in its TOML, it will be loaded and visible to buildpacks that run after it.
+- At "launch" time. When an image is finalized, if a layer is created with `launch=true`, it is TOML and will be loaded when the image is launched (such as `docker run`).
+- Inside of the same buildpack. A large buildpack may generate several layers that depend on each other. For example the `heroku/ruby` buildpack first downloads a ruby executable and places it in a `ruby` layer. It then executes code such as `gem install` and `bundle install` that depend on that ruby executable. It's a buildpack author's responsibility to make sure any environment variable modifications needed at `build` or `launch` time are also tracked and re-exported (or passed) to code running in the same buildpack.
+
+The "inside of the same buildpack" relies on the order of buildpack execution to load environment modifications, while build and launch rely on the layer's name. The more layers added to a buildpack, the less likely the three ordering locations will match exactly.
+
+# Motivation
+[motivation]: #motivation
+
+## Problem statement
+
+- Today a single buildpack can write multiple layers.
+- For launch or the next buildpack in build, each layer is evaluated in alphabetic order via the spec.
+- This order can differ from the order that layers were written by the buildpack which is surprising and can result in difficult to debug problem.
+- The current ordering mechanism requires prefixing a directory to change order. Some binaries cannot be relocated without being rebuilt.
+- We should introduce a mechanism that allows buildpack authors to ensure that build, launch, and "inside of the same buildpack" have the same order of environment variable modification. This mechanism should allow non-relocatable binaries to be re-ordered.
+
+> - Why should we do this?
+
+To improve the correctness and reliability of buildpacks.
+
+> - What use cases does it support?
+
+- Re-ordering layers without having to move files on disk.
```

Comment:
I don't yet understand why we need to re-order layers by moving files on disk?  For example, the packteo `php-composer` buildpack creates 3 layers (https://github.com/paketo-buildpacks/php-composer/blob/428c4ce4577621703065c5f5a4f5e99a7c7b4751/packages/contributor.go#L80).  If the buildpack wants to explicitly order these layers then it can, as you point out, name them `01composer`, `02php-composer-packages` and `03php-composer-cache`?